plutus-core/plutus-ir/src/PlutusIR/Core/Instance/Pretty/Readable.hs

```diff
@@ -0,0 +1,196 @@
+viewLam t@LamAbs{} = Just (go t)
+  where go (LamAbs _ n t b) = first ((n, t):) $ go b
+        go t                = ([], t)
+viewLam _          = Nothing
+
+-- | Split a let into a sequence of lets and its remaining body
+viewLet :: Term tyname name uni fun ann -> Maybe ([(Recursivity, [Binding tyname name uni fun ann])], Term tyname name uni fun ann)
+viewLet t@Let{} = Just (go t)
+  where go (Let _ r bs b) = first ((r, toList bs):) $ go b
+        go t              = ([], t)
+viewLet _       = Nothing
+
+type PrettyConstraints configName tyname name uni fun =
+  ( PrettyReadableBy configName tyname
+  , PrettyReadableBy configName name
+  , GShow uni, Closed uni, uni `Everywhere` PrettyConst, Pretty fun
+  )
+
+instance PrettyConstraints configName tyname name uni fun
+          => PrettyBy (PrettyConfigReadable configName) (Term tyname name uni fun a) where
+    prettyBy = inContextM $ \case
+        Constant _ con -> unitDocM $ pretty con
+        Builtin _ bi   -> unitDocM $ pretty bi
+        (viewApp -> Just (fun, args)) ->
+          compoundDocM juxtFixity $ \ prettyIn ->
+            let ppArg (Left a)  = braces $ prettyIn ToTheRight botFixity a
+                ppArg (Right t) = prettyIn ToTheRight juxtFixity t
+            -- Using `align` here and gathering the arguments together helps to lay out
+            -- function applications compactly like:
+            --
+            -- foo a b c
+            --     d e f
+            --
+            -- or
+            --
+            -- foo
+            --  veryLongArg
+            --  a b c
+            --
+            in prettyIn ToTheLeft juxtFixity fun <?> align (fillSep (map ppArg args))
+        Apply{}    -> error "The impossible happened. This should be covered by the `viewApp` case above."
+        TyInst{}   -> error "The impossible happened. This should be covered by the `viewApp` case above."
+        Var _ name -> prettyM name
+        (viewTyAbs -> Just (args, body)) ->
+            withPrettyAt ToTheRight botFixity $ \ prettyBot -> do
+                let pBody = prettyBot body
+                pBinds <- for args $ \(a, k) -> prettyM $ TyVarDecl () a (void k)
+                -- See comment below about laying out lambdas
+                encloseM binderFixity $ ("/\\" <> align (fillSep pBinds) <+> "->") <?> pBody
+        TyAbs{}    -> error "The impossible happened. This should be covered by the `viewTyAbs` case above."
+        (viewLam -> Just (args, body)) ->
+            -- Lay out abstraction like
+            --  \ (x : t) (y : t')
+            --    (z : t'') -> body
+            -- or
+            --  \ (x : t) (y : t')
+            --    (z : t'') ->
+            --    bigStartOfBody
+            compoundDocM binderFixity $ \prettyIn ->
+                let prettyBot x = prettyIn ToTheRight botFixity x
+                    prettyBinds = align $ fillSep [ parens (prettyBot name <+> ":" <+> prettyBot ty)
```

Comment:
This is inconsistent: you use `TyVarDecl` above, but not `VarDecl` here. Both would be simpler if the view functions returned them.